File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/viewmodel/AssetsViewModel.kt
package com.humblesolutions.indsphinx.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.humblesolutions.indsphinx.repository.BackendOccupantAssetRepository
import com.humblesolutions.indsphinx.usecase.ObserveOccupantAssetsUseCase
import com.humblesolutions.indsphinx.usecase.OccupantAssetsSnapshot
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class AssetsUiState {
    object Loading : AssetsUiState()
    data class Loaded(val snapshot: OccupantAssetsSnapshot) : AssetsUiState()
    data class Error(val message: String) : AssetsUiState()
}

class AssetsViewModel(application: Application) : AndroidViewModel(application) {
    private val repo = BackendOccupantAssetRepository()
    private val observeUseCase = ObserveOccupantAssetsUseCase(repo)

    private val _uiState = MutableStateFlow<AssetsUiState>(AssetsUiState.Loading)
    val uiState: StateFlow<AssetsUiState> = _uiState.asStateFlow()

    private var listenerJob: Job? = null
    private var listeningUid: String = ""

    /**
     * Deep-link entry can call this before the profile has loaded, so blank
     * uids are ignored: the screen re-fires once the real uid arrives. Calling
     * again with the same uid keeps the existing listener.
     */
    fun start(authUid: String) {
        if (authUid.isBlank()) return
        if (listenerJob?.isActive == true && listeningUid == authUid) return
        listenerJob?.cancel()
        listeningUid = authUid
        _uiState.value = AssetsUiState.Loading
        listenerJob = viewModelScope.launch {
            try {
                observeUseCase.execute(authUid).collect { snapshot ->
                    _uiState.value = AssetsUiState.Loaded(snapshot)
                }
            } catch (e: Exception) {
                _uiState.value = AssetsUiState.Error(e.message ?: "Could not load your assets.")
            }
        }
    }

    override fun onCleared() {
        listenerJob?.cancel()
        super.onCleared()
    }
}

File: iosApp/iosApp/AssetsViewModel.swift
import Foundation
import FirebaseFirestore

@MainActor
class AssetsViewModel: ObservableObject {
    private let repo = BackendOccupantAssetRepository()
    private var listenerRegistration: ListenerRegistration?
    private var listeningUid: String = ""

    enum State {
        case loading
        case loaded(OccupantAssetsSnapshot)
        case error(String)
    }

    @Published var state: State = .loading

    /// Deep-link entry can call this before the profile has loaded, so blank
    /// uids are ignored: the view calls again once the real uid arrives.
    /// Calling again with the same uid keeps the existing listener.
    func start(authUid: String) {
        guard !authUid.isEmpty else { return }
        if listenerRegistration != nil && listeningUid == authUid { return }
        listenerRegistration?.remove()
        listeningUid = authUid
        state = .loading
        listenerRegistration = repo.observeByAuthUid(authUid: authUid) { [weak self] snapshot in
            guard let self else { return }
            Task { @MainActor in
                self.state = .loaded(snapshot)
            }
        }
    }

    func stop() {
        listenerRegistration?.remove()
        listenerRegistration = nil
        listeningUid = ""
    }

    deinit {
        listenerRegistration?.remove()
    }
}
